### PR TITLE
Read zone information from PowerFlex secret

### PR DESCRIPTION
**Pull request by EvgenyUglov**
# Description
PR adds the method to read the zoning information from the PowerFlex array secret

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/1613|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility
- [x] I have executed the relevant end-to-end test scenarios

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] unit tests
```
# go test ./... -cover -count=1
        github.com/dell/csm-operator/api/v1             coverage: 0.0% of statements
        github.com/dell/csm-operator            coverage: 0.0% of statements
        github.com/dell/csm-operator/core/semver                coverage: 0.0% of statements
?       github.com/dell/csm-operator/pkg/constants      [no test files]
        github.com/dell/csm-operator/core               coverage: 0.0% of statements
        github.com/dell/csm-operator/pkg/logger         coverage: 0.0% of statements
        github.com/dell/csm-operator/tests/shared               coverage: 0.0% of statements
        github.com/dell/csm-operator/tests/shared/clientgoclient                coverage: 0.0% of statements
        github.com/dell/csm-operator/tests/shared/crclient              coverage: 0.0% of statements
ok      github.com/dell/csm-operator/controllers        55.024s coverage: 89.1% of statements
ok      github.com/dell/csm-operator/k8s        0.068s  coverage: 90.3% of statements
ok      github.com/dell/csm-operator/pkg/drivers        0.234s  coverage: 96.1% of statements
ok      github.com/dell/csm-operator/pkg/modules        2.153s  coverage: 89.7% of statements
ok      github.com/dell/csm-operator/pkg/resources/configmap    0.032s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/resources/csidriver    0.093s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/resources/daemonset    0.117s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/resources/deployment   0.114s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/resources/rbac 0.041s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/resources/serviceaccount       0.025s  coverage: 100.0% of statements
ok      github.com/dell/csm-operator/pkg/utils  3.110s  coverage: 81.9% of statements
```
- [x] E2E tests (PowerFlex standalone tests only, since new function isn't used anywhere yet)
image

[pflex-e2e-suite-standalone.txt](https://github.com/user-attachments/files/18027137/pflex-e2e-suite-standalone.txt)

### CODE REVIEW

File: pkg/drivers/powerflex.go [hunk 1]
```diff
@@ -346,3 +346,56 @@ func ModifyPowerflexCR(yamlString string, cr csmv1.ContainerStorageModule, fileT
...
+	arraySecret, err := utils.GetSecret(ctx, secret, namespace, kube)
+	if err != nil {
+		return nil, fmt.Errorf("reading secret [%s] error [%s]", secret, err)
+	}
+
+	type StorageArrayConfig struct {
+		SystemID string `json:"systemId"`
+		Zone     struct {
+			Label string `json:"label"`
+		} `json:"zone"`
```
Comment (ChristianAtDell) [inline pkg/drivers/powerflex.go:363]: Zones in the secret will have both the 'name' and the 'label' fields, right? Do we need to keep track of both of those?

File: pkg/drivers/powerflex_test.go [hunk 2]
```diff
@@ -194,3 +196,126 @@ func TestModifyPowerflexCR(t *testing.T) {
...
+`
+	dataWithZone := `
+- username: "admin"
+  password: "password"
+  systemID: "2b11bb111111bb1b"
+  endpoint: "https://127.0.0.2"
+  skipCertificateValidation: true
+  mdm: "10.0.0.3,10.0.0.4"
+  zone:
+    label: topology.kubernetes.io/zone=US-EAST
```
Comment (ChristianAtDell) [inline pkg/drivers/powerflex_test.go:221]: Secret should have both 'labelKey' and 'name' fields, I think-- check with @tdawe or double-check the latest proposed spec

File: pkg/drivers/powerflex.go [hunk 3]
```diff
@@ -346,3 +346,56 @@ func ModifyPowerflexCR(yamlString string, cr csmv1.ContainerStorageModule, fileT
...
+		err = yaml.Unmarshal(configs, &yamlConfig)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal multi-array configuration[%v]", err)
+		}
+
+		for _, configParam := range yamlConfig {
+			if configParam.SystemID == "" {
+				return nil, fmt.Errorf("invalid value for SystemID")
+			}
+			if configParam.Zone.Label != "" {
```
Comment (ChristianAtDell) [inline pkg/drivers/powerflex.go:385]: I think that your code is accessing the secret as if it only has one field, and that field is in the format of "<labelkey>=<zone name>"? Which is what my other two comments are referring to. But I don't believe that's the structure of the proposed zoning secrets.

File: pkg/drivers/powerflex.go (hunk 1 above)
Comment (EvgenyUglov) [reply pkg/drivers/powerflex.go]: Ahh, right, confirmed with Trevor. Was confused with a different acceptance criteria:
```
zone:
  label: <user defined label name>=<zone name>
```
Fixed. Also, fixed the AC.

File: pkg/drivers/powerflex_test.go (hunk 2 above)
Comment (EvgenyUglov) [reply pkg/drivers/powerflex_test.go]: Yes, you're right. Thanks! confirmed with Trevor. Was confused with a different acceptance criteria:
Fixed. Also, fixed the AC.

File: pkg/drivers/powerflex.go (hunk 3 above)
Comment (EvgenyUglov) [reply pkg/drivers/powerflex.go]: Fixed

**Review by ChristianAtDell — APPROVED** [review]

**Review by gallacher — APPROVED** [review]
LGTM

**Pull request merged**